### ISSUE TITLE
Minor issues integrating as app in larger projects

### ISSUE THREAD

**Issue by TomLippincott** [hltcoe/turkle#174]
@cash @charman 
Hi guys, I've finally gotten back around to this and am taking a shot at integrating it into the CDH site.  The custom user class was an easy fix (see #173) and it's very close to working, but there are three small things that I've run into: 

* Two of the migrations in `0011_add_turkle_user_admin_group.py` throw an error (or one does, so the second doesn't run).  I've put the stack trace down below.
* Turkle-related models don't show up in the overarching project's admin interface, even when logged in as a superuser+staff (though I guess it might still be a question of permissions, not up-to-speed on guardian etc).  I've tried sprinkling autodiscover in various places to no avail, and the other (but much simpler) CDH apps all show up, so I'm thinking maybe something to do with the `admin.py` customizations?
* Probably most important: when creating a batch and previewing it, the iframe displays frowny-face with "localhost refused to connect".  However, if I directly enter the URL the iframe is supposed to be showing (e.g. "TURKLE_ROOT/task/30/iframe/"), or directly click it when viewing the page source code, it renders the HIT just fine.

Any thoughts on these?  Happy to put in more pull requests, but my understanding of turkle's design is rather shallow at the moment.  Here's that migration stack trace:

```
Using selector: EpollSelector
Operations to perform:
  Apply all migrations: admin, auth, cdh, contenttypes, django_nyt, django_registration, guardian, image_clustering, matchmaking, primary_sources, schedule, sessions, sites, thumbnail, topic_modeling, turkle, wiki
Running migrations:
  Applying turkle.0011_add_turkle_user_admin_group...Traceback (most recent call last):
  File "/home/tom/cdh-domain/manage.py", line 22, in <module>
    main()
  File "/home/tom/cdh-domain/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/core/management/base.py", line 330, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/core/management/base.py", line 371, in execute
    output = self.handle(*args, **options)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/core/management/base.py", line 85, in wrapped
    res = handle_func(*args, **kwargs)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 243, in handle
    post_migrate_state = executor.migrate(
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/db/migrations/migration.py", line 124, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/home/tom/turkle/turkle/migrations/0011_add_turkle_user_admin_group.py", line 47, in add_turkle_user_admin_group
    user_perms = [Permission.objects.get(codename=codename, content_type=user_ctype)
  File "/home/tom/turkle/turkle/migrations/0011_add_turkle_user_admin_group.py", line 47, in <listcomp>
    user_perms = [Permission.objects.get(codename=codename, content_type=user_ctype)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/tom/cdh-domain/local/lib/python3.9/site-packages/django/db/models/query.py", line 429, in get
    raise self.model.DoesNotExist(
django.contrib.auth.models.Permission.DoesNotExist: Permission matching query does not exist.
```

**Comment by cash** [comment]
Thanks @TomLippincott for the report and PR. I'll test out the user PR today.

On the admin site, I think any sufficiently complicated Django site has customizations for the admin. For Turkle, look in admin.py and at the bottom, you'll see this:
```
admin_site = TurkleAdminSite(name='turkle_admin')
admin_site.register(Group, CustomGroupAdmin)
admin_site.register(User, CustomUserAdmin)
admin_site.register(Batch, BatchAdmin)
admin_site.register(Project, ProjectAdmin)
```
CDH likely has something similar and will only display the models/admins registered that way. I'm not sure what is the best Django way to handle this. I assume you don't want to modify CDH. Or the larger problem is that you may not want CDH to know about Turkle nor Turkle to know about CDH in which case you may need an integration app that handles the various registrations that need to occur.

**Comment by cash** [comment]
I think I fixed the iframe issue that Django 3 introduced here: https://github.com/hltcoe/turkle/commit/808c4d84e20db282b5d4ae786d3f1950a8a70c27

Are you running with the latest in master?

**Comment by cash** [comment]
Wait, I see that was a change I made to the turkle_site settings so you wouldn't pick that up with a pip installed version of Turkle. 

Also, I plan to deploy to PyPI soon now that Turkle is installable.

**Comment by TomLippincott** [comment]
@cash OK, that merge request is fixed, and it sounds like I should try some fiddling to get the models visible in the primary admin interface, shouldn't be a problem.  An issue I'm having trouble with though is the batch review iframe being very small (a couple hundred pixels in both dimensions, showing just a corner of the HIT with scrollbars to see the rest).  I can't figure out what about the containing/contained elements, or stylesheets/scripts, is making this happen.  Have you run into this?

**Comment by cash** [comment]
@TomLippincott I updated the migration script you pointed out as a problem to use the custom auth model. Let me know if that fixes that problem.

If it would make integration easier, I could add a function like this:
```python
def turkle_register_with_admin(site, include_user=True):
    site.register(Batch, BatchAdmin)
    ...
```
With the idea that we might add further admin models in the future and this would pick them up. You'd just have to make sure to call that and pass in your sites admin site object for an integration. First step is to confirm that the issue is registering with the admin site object and then we can discuss whether this helps.

On the display issue for batch preview, my guess is a conflict in CSS between our template and your site. I'll take a look at the key CSS rules and post them here for you to test.

**Comment by cash** [comment]
In turkle admin css, we include this set of rules:
```css
iframe {
    top: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border: 0;
    box-sizing: inherit;
    display: block;
}
```
If I turn off the width and height rules, I see this on the preview page:
![image](https://user-images.githubusercontent.com/199558/157342411-f76447a7-fa31-4b5f-ad0d-c262462631be.png)

**Comment by TomLippincott** [comment]
Ah OK, this has something to do with template inheritance, probably on my end (there are a *lot* of "base(_site)?.html" files floating around).  Thanks for the help!

**Comment by TomLippincott** [comment]
The remaining issue with admin interface integration is the hard-coded "turkle_admin" namespace in the ModelAdmin instances.  I can directly add the models to the main admin site (your suggested for loop, but without the second arguments) and they show up as expected, but the admin customizations are important for end-users creating HITs.  I might be able to hack together something by essentially duplicating then tinkering with the Turkle admin.py file, but wonder if BatchAdmin etc can be abstracted from the admin site name for direct reuse.

**Comment by cash** [comment]
@TomLippincott You are referring to the URLs, correct? 

For example, this is the stats() method for ProjectAdmin:
```
    def stats(self, obj):
        stats_url = reverse('turkle_admin:project_stats', kwargs={'project_id': obj.id})
        return format_html('<a href="{}" class="button">Stats</a>'.
                           format(stats_url))
```
The reverse() function is passed 'turkle_admin:project_stats' which is defined in the get_urls() function.

And the urls are provided with this:
```
    def get_urls(self):
        urls = super().get_urls()
        my_urls = [
            path('autocomplete-order-by-name',
                 self.admin_site.admin_view(ProjectSearchView.as_view(model_admin=self)),
                 name='autocomplete_project_order_by_name'),
            path('<int:project_id>/activity.json',
                 self.admin_site.admin_view(self.activity_json), name='project_activity_json'),
            path('<int:project_id>/stats/',
                 self.admin_site.admin_view(self.project_stats), name='project_stats'),
            path('active-projects/',
                 self.admin_site.admin_view(self.active_projects), name='active_projects'),
            path('active-users/',
                 self.admin_site.admin_view(self.active_users), name='active_users'),
        ]
        return my_urls + urls
```

I'd have to play around with this to see how to make this configurable.

**Comment by TomLippincott** [comment]
Yeah, I'm not sure what exactly the main admin interface does internally to manage models from different apps, but it must at least involve building its own urlconf where that turkle_admin namespace doesn't resolve.

**Comment by cash** [comment]
Is the issue that there are dead links on the project and batch pages or are you getting a generic model page for those?

**Comment by TomLippincott** [comment]
If I have the following in my overarching project:

```
from django.contrib import admin
from turkle.models import Batch, Project
from turkle.admin import BatchAdmin, ProjectAdmin

model_classes = [
    (Batch, BatchAdmin),
    (Project, ProjectAdmin),
]

for m, ma in model_classes:
    admin.site.register(m, ma)
```

the models show up in the admin interface, but browsing to e.g. "Batches" gives:

```
NoReverseMatch at /admin/turkle/batch/
'turkle_admin' is not a registered namespace
```
etc.  Changing to `admin.site.register(m)` makes everything work, but of course without any of the Turkle admin customizations.

**Comment by cash** [comment]
@TomLippincott I got this to work. In my site urls, I added this:

from turkle.admin import admin_site as turkle_admin_site

urlpatterns = [
    path('polls/', include('polls.urls')),
    path('admin/', admin.site.urls),
    path('turkle/', include('turkle.urls')),
    path('accounts/', include('django.contrib.auth.urls')),
    path('turkle_admin/', turkle_admin_site.urls),
]

This puts the turkle admin on a different URL than the main admin. There could be interactions between the custom user and group models that we register. I didn't get a chance to test that.

**Comment by TomLippincott** [comment]
Thanks for looking into this!  I didn't actually have a problem getting the independent Turkle admin to work (and it worked fine with my custom User model, FWIW): the issue was with having the main admin using the modeladmin customizations, which I *think* should be possible, and very nearly works, except that "turkle_admin" must be hard-coded somewhere in a way that breaks that modularity.

**Comment by cash** [comment]
@TomLippincott right - that's what I fixed by including the turkle_admin_site object and registering its urls. It's possible that you could include them directly to the `admin/` path but I didn't test that.

@charman introduced the custom site admin object in this commit: https://github.com/hltcoe/turkle/commit/da049542d37c163bfa3cb72b77b943c4def89dd2
It uses a custom name "turkle_admin" that is causing the integration issues. Another possible change is to remove that custom name. Maybe @charman remembers why he used the custom name rather than "admin".

**Comment by cash** [comment]
@TomLippincott My understanding is that the `name` of an admin site is only used in setting the nickname of URLs. This would cause a problem with adding two apps that both used the URL nickname 'admin:publish_batch` as an example. You're seeing 'turkle_admin' errors because we are creating URLs using the reverse function like this: `reverse('turkle_admin:review_batch')`

One option is changing the name of our admin site to "admin" and then adding "turkle:" to all our admin URL names. So our publish_batch url name would be 'admin:turkle:publish_batch". This would prevent any URL name conflicts with another app.

I also plan to look around at other Django apps that have admin models to see how they handle this.

**Comment by cash** [comment]
I see three options in the Django apps that I locked through:
 1. Define your URL names and hope there aren't conflicts
 2. Move your URL names into your settings so it allows a site administrator to redefine if conflict
 3. Namespace your URL strings

My plan unless I get push back is to change to using the default site name and namespace our URLs to avoid any future conflict with another app.

**Comment by TomLippincott** [comment]
@cash Looking at turkle/admin.py, what if the ModelAdmin subclasses had a self.namespace property and interpolated that into URLs rather than hard-coding?  That way, another admin interface could use them by trivially subclassing them to redefine that value appropriately.  I wouldn't be surprised if there is a "right" way to do this that leverages the fact that each ModelAdmin gets registered with the admin site(s), but it wasn't obvious to me from the docs (e.g. a programmatic way for an instance of a ModelAdmin to determine the admin site it is operating under).

I think all the other stuff in admin.py that isn't a ModelAdmin method always has *access* to an instance, so the namespace is available there, too.

**Comment by cash** [comment]
closed by #176
